python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1156,26 +1156,40 @@ class NamespacedMetadataEntries(ABC, BaseModel, frozen=True):
     def namespace(cls) -> str:
         raise NotImplementedError()
 
+    @classmethod
+    def _namespaced_key(cls, key: str) -> str:
+        return f"{cls.namespace()}/{key}"
+
+    @staticmethod
+    def _strip_namespace_from_key(key: str) -> str:
+        return key.split("/", 1)[1]
+
+    def dict(self):
+        return {self._namespaced_key(key): getattr(self, key) for key in self.__fields__.keys()}
+
     def keys(self) -> AbstractSet[str]:
-        dictionary = self.dict()
         return {
-            f"{self.namespace()}/{key}" for key in dictionary.keys() if dictionary[key] is not None
+            self._namespaced_key(key)
+            for key in self.__fields__.keys()
+            if getattr(self, key) is not None
         }
 
     def __getitem__(self, key: str) -> Any:
-        return self.dict()[key.split("/", 1)[1]]
+        return getattr(self, self._strip_namespace_from_key(key))
```

Comment:
it's not super obvious that `getattr` ends up getting the pydanic property on the derived class. Maybe add a comment?